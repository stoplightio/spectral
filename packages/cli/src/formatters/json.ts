import { Formatter } from './types';

export const json: Formatter = results => {
  const outputJson = results.map(result => {
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
