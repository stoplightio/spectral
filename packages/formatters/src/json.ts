import { Formatter } from './types';

export const json: Formatter = results => {
  const outputJson = results.map(result => {
    let documentationUrlObject = {};
    if (result.documentationUrl) {
      documentationUrlObject = {
        documentationUrl: result.documentationUrl,
      };
    }
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
      ...documentationUrlObject,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
