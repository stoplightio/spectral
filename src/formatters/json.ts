import { IRuleResult } from '../types';

export const json = (results: IRuleResult[]): string => {
  const outputJson = results.map(result => {
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
