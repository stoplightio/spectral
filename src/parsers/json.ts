import { getLocationForJsonPath, JsonParserResult, parseWithPointers, trapAccess } from '@stoplight/json';
import { IParser } from './types';

export const parseJson = (input: string) =>
  parseWithPointers(input, { ignoreDuplicateKeys: false, preserveKeyOrder: true });

export const Json: IParser<JsonParserResult<unknown>> = {
  parse: parseJson,
  getLocationForJsonPath,
  trapAccess,
};
