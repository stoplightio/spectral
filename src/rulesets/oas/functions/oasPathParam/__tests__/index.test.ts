import { RuleType, Spectral } from '../../../../../index';
import { commonOasFunctions } from '../../../index';

import { rules } from '../../../index.json';

const ruleset = { functions: commonOasFunctions(), rules };

describe('oasPathParam', () => {
  const s = new Spectral();
  s.addFunctions(ruleset.functions || {});
  s.addRules({
    'path-params': Object.assign(ruleset.rules['path-params'], {
      recommended: true,
      type: RuleType[ruleset.rules['path-params'].type],
    }),
  });

  test('No error if templated path is not used', async () => {
    const results = s.run({
      paths: {
        '/foo': {
          get: {},
        },
      },
    });
    expect(results).resolves.toHaveLength(0);
  });

  test('Error if no path parameter definition', async () => {
    const results = await s.run({
      paths: {
        '/foo/{bar}': {
          get: {},
        },
      },
    });
    expect(results).toMatchSnapshot();

    expect(results[0].path).toEqual(['paths', '/foo/{bar}']);
    expect(results[0].message)
      .toEqual(`The path "**/foo/{bar}**" uses a parameter "**{bar}**" that does not have a corresponding definition.

To fix, add a path parameter with the name "**bar**".`);
  });

  test('No error if path parameter definition is used (at the path level)', async () => {
    const results = s.run({
      paths: {
        '/foo/{bar}': {
          parameters: [
            {
              name: 'bar',
              in: 'path',
              required: true,
            },
          ],
          get: {},
        },
      },
    });
    expect(results).resolves.toHaveLength(0);
  });

  test('No error if $ref path parameter definition is used (at the path level)', async () => {
    const results = s.run({
      paths: {
        '/foo/{bar}': {
          parameters: [
            {
              $ref: '#/definitions/barParam',
            },
          ],
          get: {},
        },
      },
      definitions: {
        barParam: {
          name: 'bar',
          in: 'path',
          required: true,
        },
      },
    });
    expect(results).resolves.toHaveLength(0);
  });

  test('No error if path parameter definition is set (at the operation level)', async () => {
    const results = s.run({
      paths: {
        '/foo/{bar}': {
          get: {
            parameters: [
              {
                name: 'bar',
                in: 'path',
                required: true,
              },
            ],
          },
        },
      },
    });
    expect(results).resolves.toHaveLength(0);
  });

  test('Error if duplicate path parameters with same name are used', async () => {
    const results = await s.run({
      paths: {
        '/foo/{bar}/{bar}': {
          parameters: [
            {
              name: 'bar',
              in: 'path',
              required: true,
            },
          ],
          get: {},
        },
      },
    });
    expect(results).toMatchSnapshot();

    expect(results[0].path).toEqual(['paths', '/foo/{bar}/{bar}']);
    expect(results[0].message).toEqual(`The path "**/foo/{bar}/{bar}**" uses the parameter "**{bar}**" multiple times.

Path parameters must be unique.

To fix, update the path so that all parameter names are unique.`);
  });

  test('Error if $ref path parameter definition is not required', async () => {
    const results = await s.run({
      paths: {
        '/foo/{bar}': {
          parameters: [
            {
              $ref: '#/definitions/barParam',
            },
          ],
          get: {},
        },
      },
      definitions: {
        barParam: {
          name: 'bar',
          in: 'path',
          required: false,
        },
      },
    });
    expect(results).toMatchSnapshot();

    expect(results[0].path).toEqual(['paths', '/foo/{bar}', 'parameters']);
  });

  test('Error if paths are functionally equivalent', async () => {
    const results = await s.run({
      paths: {
        '/foo/{boo}': {
          parameters: [
            {
              name: 'boo',
              in: 'path',
              required: true,
            },
          ],
          get: {},
        },
        '/foo/{bar}': {
          parameters: [
            {
              name: 'bar',
              in: 'path',
              required: true,
            },
          ],
          get: {},
        },
      },
    });
    expect(results).toMatchSnapshot();

    expect(results[0].path).toEqual(['paths']);
  });
});
