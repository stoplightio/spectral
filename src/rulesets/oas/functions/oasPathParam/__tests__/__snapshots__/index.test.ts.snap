// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oasPathParam Error if $ref path parameter definition is not required 1`] = `
Array [
  Object {
    "code": "path-params",
    "message": "Path parameter \\"**bar**\\" must have a \`required\` that is set to \`true\`.

To fix, mark this parameter as required.",
    "path": Array [
      "paths",
      "/foo/{bar}",
      "parameters",
    ],
    "range": Object {
      "end": Object {
        "character": 42,
        "line": 5,
      },
      "start": Object {
        "character": 19,
        "line": 3,
      },
    },
    "severity": 0,
    "source": undefined,
    "summary": "Path parameter \\"**bar**\\" must have a \`required\` that is set to \`true\`.

To fix, mark this parameter as required.",
  },
]
`;

exports[`oasPathParam Error if duplicate path parameters with same name are used 1`] = `
Array [
  Object {
    "code": "path-params",
    "message": "The path \\"**/foo/{bar}/{bar}**\\" uses the parameter \\"**{bar}**\\" multiple times.

Path parameters must be unique.

To fix, update the path so that all parameter names are unique.",
    "path": Array [
      "paths",
      "/foo/{bar}/{bar}",
    ],
    "range": Object {
      "end": Object {
        "character": 15,
        "line": 10,
      },
      "start": Object {
        "character": 23,
        "line": 2,
      },
    },
    "severity": 0,
    "source": undefined,
    "summary": "The path \\"**/foo/{bar}/{bar}**\\" uses the parameter \\"**{bar}**\\" multiple times.

Path parameters must be unique.

To fix, update the path so that all parameter names are unique.",
  },
]
`;

exports[`oasPathParam Error if no path parameter definition 1`] = `
Array [
  Object {
    "code": "path-params",
    "message": "The path \\"**/foo/{bar}**\\" uses a parameter \\"**{bar}**\\" that does not have a corresponding definition.

To fix, add a path parameter with the name \\"**bar**\\".",
    "path": Array [
      "paths",
      "/foo/{bar}",
    ],
    "range": Object {
      "end": Object {
        "character": 15,
        "line": 3,
      },
      "start": Object {
        "character": 17,
        "line": 2,
      },
    },
    "severity": 0,
    "source": undefined,
    "summary": "The path \\"**/foo/{bar}**\\" uses a parameter \\"**{bar}**\\" that does not have a corresponding definition.

To fix, add a path parameter with the name \\"**bar**\\".",
  },
]
`;

exports[`oasPathParam Error if paths are functionally equivalent 1`] = `
Array [
  Object {
    "code": "path-params",
    "message": "The paths \\"**/foo/{boo}**\\" and \\"**/foo/{bar}**\\" are equivalent.

To fix, remove one of the paths or merge them together.",
    "path": Array [
      "paths",
    ],
    "range": Object {
      "end": Object {
        "character": 15,
        "line": 20,
      },
      "start": Object {
        "character": 10,
        "line": 1,
      },
    },
    "severity": 0,
    "source": undefined,
    "summary": "The paths \\"**/foo/{boo}**\\" and \\"**/foo/{bar}**\\" are equivalent.

To fix, remove one of the paths or merge them together.",
  },
]
`;
